### PR TITLE
Add connectionist temporal classification (CTC) loss algorithm

### PR DESCRIPTION
**Pull request by ZJUGuoShuai**
### Describe your change:

- [Connectionist Temporal Classification (CTC)](https://en.wikipedia.org/wiki/Connectionist_temporal_classification) loss is used in speech recognition, handwriting recognition and other sequence problems. It's used to get around not knowing the alignment between the input and the output.
- The implementation has been verified to align with PyTorch's [CTCLoss](https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html).

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

Fixes #13410

### CODE REVIEW

**Review by imSanko — APPROVED** [review]

**Comment by AtomicVar** [comment]
Hi there, I just wanted to thank the maintainers for their hard work on the project, and I wanted to let them know that I submitted a few pull requests in the last few days. I would really appreciate it if one of them could take a look at them and let me know if there are any issues or if there's anything that needs to be fixed before they can be merged.

Thanks again for all your hard work! @cclauss

File: machine_learning/loss_functions.py [hunk 1]
```diff
@@ -471,6 +471,110 @@ def perplexity_loss(
...
+    Calculate the connectionist temporal classification (CTC) loss between the given
+    log probabilities and targets.
+
+    CTC loss is used in speech recognition, handwriting recognition and other sequence
+    problems. It's used to get around not knowing the alignment between the input and
+    the output.
+
+    References:
+    - https://en.wikipedia.org/wiki/Connectionist_temporal_classification
+    - https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html
```
Comment (tianyizheng02) [inline machine_learning/loss_functions.py:487]: A couple suggestions to clarify the documentation:
1. PyTorch cites [Graves et al](https://www.cs.toronto.edu/~graves/icml_2006.pdf) for its implementation. Since your implementation is also based on this paper, please add it as a reference as well.
2. Please add a short paragraph explaining how the loss is actually calculated. This should explain some general questions about your variables (What is `blank`? What is `alpha`) so that the reader understands _what_ is being calculated. I ask for this because this repository is meant for educational purposes, so we want readers to understand _how_ and _why_ the implementation works.

    Also, in your implementation you use `np.logaddexp` for log-probabilities when calculating `alpha` rather than calculating probabilities directly. Since this differs from the definitions in the Graves et al paper, please be sure to note this implementation detail in your explanation.

**Review by tianyizheng02 — DISMISSED** [review]

**Review by cclauss — APPROVED** [review]
